Add zero padding option to `tf.image.extract_glimpse`

yongtang opened the pull request:
There has been several feature request in #16663 and https://github.com/tensorflow/tensorflow/issues/2134#issuecomment-326071624 asking for zero padding option to `tf.image.extract_glimpse`.

This fix is an attempt to address the issue through:
1. Add an additional args of `noise` so that it is possible to specify `uniform`, `gaussian`, or `zero`.
2. In case `zero` is specified, extract_glimpse will fill padding with zeros.

Note this PR keeps the original arg `uniform_noise` intact (unless it conflict with `noise`), though it is possible to deprecate the `uniform_noise` if desired.

This fix fixes #16663.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add zero padding option to `tf.image.extract_glimpse`

There has been several feature request in 16663 and 2134
asking for zero padding option to `tf.image.extract_glimpse`.
This fix is an attempt to address the issue through:
1. Add an additional args of `noise` so that it is possible
to specify `uniform`, `gaussian`, or `zero`.
2. In case `zero` is specified, extract_glimpse will fill
padding with zeros.

Note this PR keeps the original arg `uniform_noise` intact
(unless it conflict with `noise`), though it is possible
to deprecate the `uniform_noise` if desired.

This fix fixes 16663.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update tensorflow/core/ops/image_ops.cc

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test case for additional arg with extract_glimpse

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add additional test cases

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update goldens API

with
```
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test
           --update_goldens True
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update the api_def

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Note: Not sure if it is better to have `ExtractGlimpseV2` instead?

benoitsteiner commented on tensorflow/core/kernels/attention_ops.cc:
```diff
@@ -98,13 +108,13 @@ class ExtractGlimpseOp : public OpKernel {
   bool normalized_;
   bool centered_;
-  bool uniform_noise_;
+  string noise_;
```
Can you create an enum to encode the valid types ? String comparisons are expensive, and don't run well on GPU.

yongtang replied:
@benoitsteiner Thanks. Done.

benoitsteiner commented on tensorflow/core/kernels/eigen_attention.h:
```diff
@@ -144,8 +144,11 @@ struct GlimpseExtractionOp {
-          // Initialize the glimpse with uniform noise.
+        if (noise_ == "zero") {
+           // Initialize the glimpse with zero noise.
+          output.template chip<3>(i).setZero();
```
Can you specify the device to preserve the GPU compatibility ? You can call output.template chip<3>(i).device().setZero();, or if that's not implemented something like output.template chip<3>(i) = output.template chip<3>(i).constant(0);

yongtang replied:
Thanks @benoitsteiner. Updated.

yongtang committed:
Preserve the GPU compatibility

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Convert string to enum for performance/GPU

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @benoitsteiner for the review. The PR has been updated. Please take a look.

tensorflowbutler commented:
Nagging Assignee @benoitsteiner: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Reviewer @benoitsteiner: It has been 14 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

tensorflowbutler commented:
Nagging Reviewer @benoitsteiner: It has been 14 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

tensorflowbutler commented:
Nagging Reviewer @benoitsteiner: It has been 64 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

tensorflowbutler commented:
Nagging Reviewer @benoitsteiner: It has been 79 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

tensorflowbutler commented:
Nagging Reviewer @benoitsteiner: It has been 94 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

tensorflowbutler commented:
Nagging Assignee @benoitsteiner: It has been 130 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

karmel commented:
@benoitsteiner -- Can you please take a look at the updates from @yongtang  and approve if they LGTY?

tensorflowbutler commented:
Nagging Assignee @benoitsteiner: It has been 20 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

rmlarsen commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -535,6 +536,14 @@ REGISTER_OP("ExtractGlimpse")
+      string noise;
+      TF_RETURN_IF_ERROR(c->GetAttr("noise", &noise));
+      if (uniform_noise && noise != "") {
+        return errors::InvalidArgument("The uniform_noise and noise could not be specified at the same time");
```
nit: s/could/should/

rmlarsen commented:
Please resolve conflicts.

yongtang committed:
Address review comment

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update API golden

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix `Experimental clang-format Check` failure

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@rmlarsen Thanks for the review. The PR has been rebased and with comments addressed. Please take a look and let me know if there are any other issues.

asimshankar commented:
(For API review)

Can't this break callers that use only positional arguments (since a new argument was inserted before the last one, `name`)? If so, let's add a wrapper (just for v1) to prevent that, i.e., where the new argument is after `name` in the Python API.

yongtang committed:
Add wrapper to tf.image.extract_glimpse to maintain backward-compatibility

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

asimshankar commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -2805,3 +2805,26 @@ def sobel_edges(image):
   return output
+
+
+@tf_export('image.extract_glimpse')
```
Would it make sense to make this:

```python
@tf_export(v1=['image.extract_glimpse'], v2=[])
def ...
```

And add another function that just wraps over `gen_image_ops.extract_glimpse` to be exported for v2?

yongtang committed:
Use @tf_export(v1=[..], v2=[..]) to expose image.extract_glimpse

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update api goldens

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Pylint fix

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@asimshankar Thanks for the review. The PR has been updated so that v1 and v2 exposed differently. Please take a look and let me know if there are any issues.

asimshankar commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -2805,3 +2805,29 @@ def sobel_edges(image):
+extract_glimpse.__doc__ = gen_image_ops.extract_glimpse.__doc__
+
+extract_glimpse_v2 = gen_image_ops.extract_glimpse
+tf_export(v2=['image.extract_glimpse'])(extract_glimpse_v2)
```
Ah, sorry, it's supposed to be:

```python
tf_export('image.extract_glimpse', v1=[])
```

(i.e., no `v2`)

CC @av8ramit

yongtang committed:
Fix exposed namespace api to no use v2, only `tf_export('image.extract_glimpse', v1=[])`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix api golden

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

rmlarsen commented:
@yongtang sorry for the delay. Could you please resolve the conflicts?

yongtang commented:
Thanks @rmlarsen, the PR has been rebased and updated. Please take a look.

martinwicke commented:
Thank you for pinging this. I agree with you on the 2.0 signatures. Can you change the default for noise to explicitly be "uniform" (since this is the default, right?) 

Otherwise it looks good to me. If you could update the PR that would be great.

@tensorflow/api-owners FYI.

yongtang committed:
Update image.extract_glimpse to take noise in v2, and keep v1 api

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update test cases for tf.image.extract_glimpse

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update API goldens with v2 change only

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @martinwicke, the PR has been updated as discussed.

The `Sanity` is failing though it is unrelated (due to swift build):
```
ERROR: error loading package 'tensorflow/lite/experimental/swift': Extension file not found. Unable to load file '@bazel_skylib//lib:new_sets.bzl': file doesn't exist or isn't a file
```

martinwicke commented:
@saeta do you know about this?

martinwicke reviewed (changes requested):
Can you add a transformation to tools/compatibility/tf_convert_v2.py which transforms the arguments? This would be a custom transformation, you can look at the _dropout_transformer for an example.

yongtang committed:
Set default noise as uniform to conform with v1 api

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add v2 transform compatibility tool for tf.image.extract_glimpse

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented on tensorflow/core/api_def/base_api/api_def_ExtractGlimpse.pbtxt:
```diff
@@ -48,6 +48,14 @@ END
+    description: <<END
+indicates if the noise should `uniform`, `gaussian`, or
+`zero`. The default is `` which means the the noise type
+will be decided by `uniform_noise`.
```
Thanks. Updated.

yongtang commented:
Thanks @martinwicke for the review. I have added the transformation of `extract_glimpse` in ` tools/compatibility/`, and updated the PR.

Please take a look and let me know if there are any issues.

martinwicke commented on tensorflow/tools/compatibility/tf_upgrade_v2.py:
```diff
@@ -1177,6 +1177,53 @@ def __init__(self):
+
+    # Maybe it was a positional arg
+    if len(node.args) < 5:
+      errors.append((node.lineno, node.col_offset,
```
I don't think this should be an error, right? It's legal to omit the arg, so if it's just not there, we don't have to do anything.

yongtang replied:
Thanks. Fixed.

martinwicke commented on tensorflow/tools/compatibility/tf_upgrade_v2.py:
```diff
@@ -1177,6 +1177,53 @@ def __init__(self):
+      # Put parentheses around noise.value (and remove the old prefix/
+      # suffix, they should only be around new_value).
+      pasta.base.formatting.set(new_value, "prefix", "(")
+      pasta.base.formatting.set(new_value, "suffix", ")")
```
You need these around new_value.test, not new_value, since you're making something that looks like `"uniform if blah else "gaussian"`, and if `blah` is something like `a if b else c + d`, then the resulting expression is `"uniform" if a if b else c else "gaussian"` which is a syntax error, but `"uniform" if (a if b else c) else "gaussian"` is valid.

yongtang replied:
Done.

martinwicke reviewed (changes requested):
Some details in the transformer, otherwise looks good, thank you!

yongtang committed:
Add parentheses around new_value.test to cover `d if (a if t else b) else d`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @martinwicke for the review. The PR has been updated. Please take a look.

yongtang committed:
Fixed Experimental clang-format Check error

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yifeif commented:
@martinwicke yep!

yongtang commented:
@martinwicke It looks like there is a merge conflict. I am wondering if I should resolve the merge conflict and update the PR, or wait for the `import/copybara` to complete?

martinwicke commented:
Yes, please resolve the conflict. The import won't succeed while it is
there.

>

yongtang commented:
Thanks @martinwicke. The PR has been rebased with conflict resolved.
